### DISCUSSION TITLE
JCache vs Redisson for Redis-based Bucket4J

### DISCUSSION THREAD

**Discussion by daichi-m** [vladimir-bukhtoyarov/bucket4j#445] [Q&A] [answered]
We are currently using Bucket4J to do approximate rate-limiting in our application (we do not have stringent rate-limits in place and ocassionally allow users to overuse a certain amount). It has been implemented with JCache integrated with Redis under-the-hood. I saw that recently you have also released a direct Redisson based integration for accessing the buckets. I have a few questions around these:

1. Which implementation, JCache or Redisson, is more stable in terms of consistency and latency (we do have stringent latency requirements)?
2. Does the JCache implementation sync back to the backend store immediately after tokens are consumed, and can that be used in a distributed system effectively?
3. In Redisson, I see there is an option to set TTL for the buckets. Is there a similar API in JCache too?

Help would be much appreciated.

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @daichi-m 

1. I have never heard that Bucket4j works correctly in case of JCache over Redisson. I strongly recommend to use direct integration with Redisson.
2. Everyone Bucket4j integration works correctly in distrubited scenario. By default token consumption is being synchronized imediatelly, if you want to avoid imediate synchronization you can configure some optimization during bucket construction via builder, see [this](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L64) and [this](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/optimization/Optimizations.java#L35).
3. TTL can be configured by configuring ExpirationAfterWriteStrategy [see this exampl](https://bucket4j.com/8.8.0/toc.html#example-of-bucket-instantiation-via-redissonbasedproxymanager)

**Reply by daichi-m** [reply]
Thank you @vladimir-bukhtoyarov
